crates/walrus-sui/src/client/retry_client/download_handler.rs

```diff
@@ -0,0 +1,599 @@
+            self.last_tracked_failed_checkpoint
+                .store(0, Ordering::Relaxed);
+            self.current_consecutive_failures_for_tracked_checkpoint
+                .store(0, Ordering::Relaxed);
+            tracing::debug!(
+                seq = sequence_number,
+                "Reset consecutive failure tracking after fallback success."
+            );
+        }
+    }
+
+    /// Determines the initial action for fetching a checkpoint
+    pub fn download_checkpoint_decision(&self) -> CheckpointDownloadDecision {
+        if self.config.is_none() {
+            return CheckpointDownloadDecision::AttemptPrimary;
+        }
+        if Instant::now() >= self.skip_primary_until.load(Ordering::Relaxed) {
+            CheckpointDownloadDecision::AttemptPrimary
+        } else {
+            tracing::debug!(
+                "Skipping primary RPC (overall policy) until {:?}, attempting fallback directly.",
+                self.skip_primary_until.load(Ordering::Relaxed)
+            );
+            CheckpointDownloadDecision::TryFallback(None)
+        }
+    }
+
+    /// Updates all relevant failure counters based on the current failure
+    /// and returns a snapshot of the counts relevant to the decision.
+    /// Prioritizes tracking failures for smaller sequence numbers.
+    pub fn update_and_get_failure_stats(&self, sequence_number: u64) -> CheckpointFailureStats {
+        let overall_failures_count = self.num_failures.fetch_add(1, Ordering::Relaxed) + 1;
+
+        let (consecutive_failures_for_this_seq, is_this_sequence_now_tracked) = loop {
+            let current_tracked_seq = self.last_tracked_failed_checkpoint.load(Ordering::Acquire);
+
+            if current_tracked_seq == sequence_number {
+                // This sequence is already being tracked. Increment its consecutive failure count.
+                break (
+                    self.current_consecutive_failures_for_tracked_checkpoint
+                        .fetch_add(1, Ordering::Relaxed)
+                        + 1,
+                    true,
+                );
+            } else if current_tracked_seq == 0 {
+                // No sequence is currently being tracked.
+                // We will try to start tracking this sequence using CAS
+                if self
+                    .last_tracked_failed_checkpoint
+                    .compare_exchange(0, sequence_number, Ordering::SeqCst, Ordering::Relaxed)
+                    .is_ok()
+                {
+                    // CAS succeeded and we're now tracking this sequence
+                    self.current_consecutive_failures_for_tracked_checkpoint
+                        .store(1, Ordering::Relaxed);
+                    break (1, true);
+                } else {
+                    // CAS failed as another thread started tracking a sequence.
+                    // We will loop to re-evaluate with the new state
+                    continue;
```

Comment:
I'm not 100% certain, but it feels like you can just return `(0, false)` here? Unless maybe there are other places that are updating `last_tracked_failed_checkpoint`? Seems unlikely since we know that `last_tracked_failed_checkpoint` just changed from `0` to something.. seems unlikely that it'd go straight back to `0` after a few nanoseconds, since (I assume) there are network requests involved here...